m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -504,68 +462,108 @@ public void run() {
                     // catch block.
-                    SizedMetric sizedMetric = metricQueue.poll(maxProcessorWaitUntilFlushMillis, TimeUnit.MILLISECONDS);
+                    SizedMetric sizedMetric = metricQueue.poll(maxBufferingDelay.toMillis(), TimeUnit.MILLISECONDS);
 
                     // Drop metrics that came in after close
                     if (sizedMetric == SizedMetric.CLOSE) {
                         metricQueue.clear();
                         break;
                     }
 
-                    if (sizedMetric != null) {
-                        flushMetricIteration(sizedMetric);
-                        continue;
+                    if (sizedMetric == null) {
+                        // If we didn't get any new metrics after waiting the specified time,
+                        // flush what we have so far.
+                        process(SizedMetric.FLUSH);
+                    } else {
+                        process(sizedMetric);
                     }
 
-                    // If we don't get any new metrics after waiting the specified time,
-                    // flush what we have so far.
-                    if (!pendingMetrics.isEmpty()) {
-                        flushMetricIteration(SizedMetric.FLUSH);
-                    }
                 }
             } catch (InterruptedException e) {
                 // Don't care if we get interrupted - the finally block will clean up
             } finally {
-                flushRemainingMetrics();
-
-                // Always arrive at phaser to prevent deadlock
-                phaser.arrive();
+                drainQueue();
+                flushBuffered();
+                // Count down shutdown latch to notify reporter
+                shutdownLatch.countDown();
             }
         }
 
-        private void flushMetricIteration(SizedMetric sizedMetric) {
-            Metric metric = sizedMetric.getMetric();
-
+        private void process(SizedMetric sizedMetric) {
             if (sizedMetric == SizedMetric.FLUSH) {
-                flush(pendingMetrics, commonTags);
+                flushBuffered();
                 return;
             }
 
             int size = sizedMetric.getSize();
-
-            if (metricsSize + size > freeBytes) {
-                flush(pendingMetrics, commonTags);
-                metricsSize = 0;
+            if (metricsSize + size > payloadCapacity || elapsedMaxDelaySinceLastFlush()) {
+                flushBuffered();
             }
 
-            pendingMetrics.add(metric);
+            Metric metric = sizedMetric.getMetric();
 
+            metricsBuffer.add(metric);
             metricsSize += size;
         }
 
-        private void flushRemainingMetrics() {
+        private boolean elapsedMaxDelaySinceLastFlush() {
+            return lastBufferFlushTimestamp.plus(maxBufferingDelay.toMillis(), ChronoUnit.MILLIS)
+                    .isBefore(Instant.now(clock));
```

Comment:
shouldn't this be `.isAfter`?

Reply:
It'd be `.isBefore` -- we're checking that timestamp + delay is before now